beacon-chain/db/kv/blocks.go

```diff
@@ -227,17 +231,121 @@ func (s *Store) DeleteBlock(ctx context.Context, root [32]byte) error {
 			return ErrDeleteJustifiedAndFinalized
 		}
 
-		if err := tx.Bucket(blocksBucket).Delete(root[:]); err != nil {
-			return err
-		}
-		if err := tx.Bucket(blockParentRootIndicesBucket).Delete(root[:]); err != nil {
+		if err := s.deleteBlock(tx, root[:]); err != nil {
 			return err
 		}
 		s.blockCache.Del(string(root[:]))
 		return nil
 	})
 }
 
+// DeleteHistoricalDataBeforeSlot deletes all blocks and states before the given slot.
+func (s *Store) DeleteHistoricalDataBeforeSlot(ctx context.Context, cutoffSlot primitives.Slot) error {
+	ctx, span := trace.StartSpan(ctx, "BeaconDB.DeleteHistoricalDataBeforeSlot")
+	defer span.End()
+
+	// Track validator keys that need to be deleted from cache
+	var validatorKeysToDelete [][]byte
+
+	// Perform all deletions in a single transaction for atomicity
+	return s.db.Update(func(tx *bolt.Tx) error {
+		filter := filters.NewFilter().SetStartSlot(0).SetEndSlot(cutoffSlot)
+		rootsBySlot, err := blockRootsByFilter(ctx, tx, filter)
+		if err != nil {
+			return errors.Wrap(err, "could not retrieve block roots by filter")
+		}
+
+		for slot, roots := range rootsBySlot {
+			for _, root := range roots {
+				// Delete block
+				if err = s.deleteBlock(tx, root); err != nil {
+					return err
+				}
+
+				// Delete finalized block roots index
+				if err = tx.Bucket(finalizedBlockRootsIndexBucket).Delete(root); err != nil {
+					return errors.Wrap(err, "could not delete finalized block root index")
+				}
+
+				// Delete state
+				if err = tx.Bucket(stateBucket).Delete(root); err != nil {
+					return errors.Wrap(err, "could not delete state")
+				}
+
+				// Delete state summary
+				if err = tx.Bucket(stateSummaryBucket).Delete(root); err != nil {
+					return errors.Wrap(err, "could not delete state summary")
+				}
+
+				// Clean up validator entries if state validator migration is complete
+				ok, err := s.isStateValidatorMigrationOver()
+				if err != nil {
+					return err
+				}
+				if !ok {
+					continue
+				}
+				valBkt := tx.Bucket(stateValidatorsBucket)
+				valIdxBkt := tx.Bucket(blockRootValidatorHashesBucket)
+
+				// Get validator hashes for this state
+				compressedValidatorHashes := valIdxBkt.Get(root)
+				if len(compressedValidatorHashes) > 0 {
+					validatorHashes, err := snappy.Decode(nil, compressedValidatorHashes)
+					if err != nil {
+						return errors.Wrap(err, "failed to decompress validator hashes")
+					}
+					if len(validatorHashes)%hashLength != 0 {
+						return errors.Errorf("invalid validator keys length: %d", len(validatorHashes))
+					}
+					// Delete validator entries from DB and collect keys for cache deletion
+					for i := 0; i < len(validatorHashes); i += hashLength {
+						key := validatorHashes[i : i+hashLength]
+						if err = valBkt.Delete(key); err != nil {
+							return errors.Wrap(err, "could not delete validator")
+						}
+						// Make a copy of the key before storing it
```

Comment:
Is this necessary? Since you create a new variable each time inside the loop, there shouldn't be a need for copy.

Reply:
Removed it, not needed now.